File: app/src/main/java/com/example/mcpodcasts/data/repository/PodcastRepository.kt
package com.example.mcpodcasts.data.repository

import android.content.Context
import androidx.room.withTransaction
import com.example.mcpodcasts.R
import com.example.mcpodcasts.data.local.CalendarEpisode
import com.example.mcpodcasts.data.local.EpisodeDao
import com.example.mcpodcasts.data.local.EpisodeEntity
import com.example.mcpodcasts.data.local.PodcastDao
import com.example.mcpodcasts.data.local.PodcastDatabase
import com.example.mcpodcasts.data.local.PodcastEntity
import com.example.mcpodcasts.data.local.QueueEpisode
import com.example.mcpodcasts.data.local.SubscriptionSummary
import com.example.mcpodcasts.data.rss.ParsedEpisode
import com.example.mcpodcasts.data.rss.RssFeedParser
import com.example.mcpodcasts.domain.mergeWithExistingEpisode
import com.example.mcpodcasts.domain.stableId
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.withContext
import okhttp3.OkHttpClient
import okhttp3.Request

data class NewEpisodesSummary(
    val feedUrl: String,
    val podcastTitle: String,
    val imageUrl: String?,
    val episodeTitles: List<String>,
)

class PodcastRepository(
    context: Context,
    private val database: PodcastDatabase,
    private val podcastDao: PodcastDao,
    private val episodeDao: EpisodeDao,
    private val httpClient: OkHttpClient,
    private val rssFeedParser: RssFeedParser,
) {
    private val appContext = context.applicationContext

    fun observeQueue(): Flow<List<QueueEpisode>> = episodeDao.observeQueue()

    fun observeSubscriptions(): Flow<List<SubscriptionSummary>> = podcastDao.observeSubscriptions()

    fun observeCalendarEpisodes(): Flow<List<CalendarEpisode>> = episodeDao.observeCalendarEpisodes()

    suspend fun getQueueSnapshot(): List<QueueEpisode> = withContext(Dispatchers.IO) {
        episodeDao.getQueueSnapshot()
    }

    suspend fun addSubscription(feedUrl: String): Result<Unit> = runCatching {
        val normalizedUrl = normalizeFeedUrl(feedUrl)
        syncFeed(
            feedUrl = normalizedUrl,
            notifyForNewEpisodes = false,
        )
    }

    suspend fun refreshAllFeeds(): List<NewEpisodesSummary> = withContext(Dispatchers.IO) {
        buildList {
            podcastDao.getSubscribedFeedUrls().forEach { feedUrl ->
                val syncResult = runCatching {
                    syncFeed(
                        feedUrl = feedUrl,
                        notifyForNewEpisodes = true,
                    )
                }.getOrNull() ?: return@forEach

                if (syncResult.shouldNotify && syncResult.newEpisodeTitles.isNotEmpty()) {
                    add(
                        NewEpisodesSummary(
                            feedUrl = syncResult.feedUrl,
                            podcastTitle = syncResult.podcastTitle,
                            imageUrl = syncResult.imageUrl,
                            episodeTitles = syncResult.newEpisodeTitles,
                        )
                    )
                }
            }
        }
    }

    suspend fun updatePlaybackState(
        episodeId: String,
        positionMs: Long,
        durationMs: Long,
        isRead: Boolean,
        isCompleted: Boolean,
    ) = withContext(Dispatchers.IO) {
        episodeDao.updatePlaybackState(
            episodeId = episodeId,
            positionMs = positionMs.coerceAtLeast(0L),
            durationMs = durationMs.coerceAtLeast(0L),
            isRead = isRead,
            isCompleted = isCompleted,
        )
    }

    suspend fun markEpisodeRead(
        episodeId: String,
        isRead: Boolean,
    ) = withContext(Dispatchers.IO) {
        episodeDao.markEpisodeRead(
            episodeId = episodeId,
            isRead = isRead,
        )
    }

    suspend fun markAllEpisodesReadForPodcast(
        feedUrl: String,
        isRead: Boolean,
    ) = withContext(Dispatchers.IO) {
        episodeDao.markAllEpisodesReadForPodcast(
            podcastId = feedUrl,
            isRead = isRead,
        )
    }

    suspend fun updateSubscriptionSettings(
        feedUrl: String,
        notifyNewEpisodes: Boolean,
        includeInQueue: Boolean,
        introSkipSeconds: Int,
        outroSkipSeconds: Int,
    ) = withContext(Dispatchers.IO) {
        podcastDao.updateSubscriptionSettings(
            feedUrl = feedUrl,
            notifyNewEpisodes = notifyNewEpisodes,
            includeInQueue = includeInQueue,
            introSkipSeconds = introSkipSeconds.coerceAtLeast(0),
            outroSkipSeconds = outroSkipSeconds.coerceAtLeast(0),
        )
    }

    suspend fun removeSubscription(feedUrl: String) = withContext(Dispatchers.IO) {
        database.withTransaction {
            episodeDao.deleteEpisodesForPodcast(feedUrl)
            podcastDao.deletePodcast(feedUrl)
        }
    }

    private suspend fun syncFeed(
        feedUrl: String,
        notifyForNewEpisodes: Boolean,
    ): FeedSyncResult = withContext(Dispatchers.IO) {
        val request = Request.Builder()
            .url(feedUrl)
            .header("User-Agent", "Podcast MC/1.0")
            .build()

        val response = httpClient.newCall(request).execute()
        if (!response.isSuccessful) {
            response.close()
            error(appContext.getString(R.string.error_fetch_feed, response.code))
        }

        val parsedFeed = response.use { networkResponse ->
            val body = networkResponse.body ?: error(appContext.getString(R.string.error_feed_no_content))
            body.byteStream().use(rssFeedParser::parse)
        }

        val syncedAt = System.currentTimeMillis()
        val existingPodcast = podcastDao.getPodcast(feedUrl)
        val subscribedAt = existingPodcast?.subscribedAt ?: syncedAt
        val existingEpisodes = episodeDao.getEpisodesForPodcast(feedUrl)
            .associateBy(EpisodeEntity::episodeId)
        val podcast = PodcastEntity(
            feedUrl = feedUrl,
            title = parsedFeed.title,
            author = parsedFeed.author,
            description = parsedFeed.description,
            imageUrl = parsedFeed.imageUrl,
            siteUrl = parsedFeed.siteUrl,
            subscribedAt = subscribedAt,
            lastSyncedAt = syncedAt,
            notifyNewEpisodes = existingPodcast?.notifyNewEpisodes ?: true,
            includeInQueue = existingPodcast?.includeInQueue ?: true,
            introSkipSeconds = existingPodcast?.introSkipSeconds ?: 0,
            outroSkipSeconds = existingPodcast?.outroSkipSeconds ?: 0,
        )

        val episodes = parsedFeed.episodes
            .map { parsedEpisode ->
                parsedEpisode.mergeWithExistingEpisode(
                    podcastId = feedUrl,
                    fallbackArtworkUrl = parsedFeed.imageUrl,
                    existingEpisode = existingEpisodes[parsedEpisode.stableId()],
                )
            }
        val newEpisodeTitles = episodes
            .filter { episode -> existingEpisodes[episode.episodeId] == null }
            .map(EpisodeEntity::title)

        database.withTransaction {
            podcastDao.upsertPodcast(podcast)

            if (episodes.isEmpty()) {
                episodeDao.deleteEpisodesForPodcast(feedUrl)
            } else {
                episodeDao.upsertEpisodes(episodes)
                episodeDao.deleteEpisodesNotIn(feedUrl, episodes.map(EpisodeEntity::episodeId))
            }
        }

        FeedSyncResult(
            feedUrl = feedUrl,
            podcastTitle = podcast.title,
            imageUrl = podcast.imageUrl,
            newEpisodeTitles = if (existingPodcast == null || !notifyForNewEpisodes) {
                emptyList()
            } else {
                newEpisodeTitles
            },
            shouldNotify = notifyForNewEpisodes && podcast.notifyNewEpisodes,
        )
    }

    private fun normalizeFeedUrl(feedUrl: String): String {
        val trimmed = feedUrl.trim()
        if (trimmed.isBlank()) {
            error(appContext.getString(R.string.error_invalid_rss_url))
        }

        return trimmed
            .replace("feed://", "https://")
            .replace("http://", "http://")
    }

}

private data class FeedSyncResult(
    val feedUrl: String,
    val podcastTitle: String,
    val imageUrl: String?,
    val newEpisodeTitles: List<String>,
    val shouldNotify: Boolean,
)

File: web/js/podcastRepository.js
function normalizeFeedUrl(feedUrl) {
  const trimmed = String(feedUrl || "").trim();
  if (!trimmed) {
    throw new Error("Enter a valid RSS feed URL.");
  }
  return trimmed.replace(/^feed:\/\//i, "https://");
}

function compareByPublishedAndTitleDesc(a, b) {
  if (b.publishedAt !== a.publishedAt) {
    return b.publishedAt - a.publishedAt;
  }
  return String(a.title || "").localeCompare(String(b.title || ""), undefined, { sensitivity: "base" });
}

function compareByPublishedAndTitleAsc(a, b) {
  if (a.publishedAt !== b.publishedAt) {
    return a.publishedAt - b.publishedAt;
  }
  return String(a.title || "").localeCompare(String(b.title || ""), undefined, { sensitivity: "base" });
}

function buildNewEpisodesSummary(feedUrl, podcast, newEpisodeTitles, shouldNotify) {
  return {
    feedUrl,
    podcastTitle: podcast.title,
    imageUrl: podcast.imageUrl,
    episodeTitles: shouldNotify ? newEpisodeTitles : [],
  };
}

export function createPodcastRepository({ db, rules, rssParser }) {
  const listeners = new Set();

  function subscribe(listener) {
    listeners.add(listener);
    return () => listeners.delete(listener);
  }

  function emit() {
    listeners.forEach((listener) => listener());
  }

  async function getPodcast(feedUrl) {
    return db.get("podcasts", feedUrl);
  }

  async function getEpisodesForPodcast(podcastId) {
    return db.getManyByIndex("episodes", "by_podcast_id", podcastId);
  }

  async function fetchFeed(feedUrl) {
    const proxiedUrl = `/proxy?url=${encodeURIComponent(feedUrl)}`;
    const response = await fetch(proxiedUrl, {
      headers: {
        "User-Agent": "Podcast MC/1.0",
      },
    });
    if (!response.ok) {
      throw new Error(`Couldn't fetch the feed (${response.status}).`);
    }
    return response.text();
  }

  async function syncFeed(feedUrl, notifyForNewEpisodes) {
    const normalizedUrl = normalizeFeedUrl(feedUrl);
    const xml = await fetchFeed(normalizedUrl);
    const parsedFeed = rssParser.parse(xml);
    const syncedAt = Date.now();

    const existingPodcast = await getPodcast(normalizedUrl);
    const subscribedAt = existingPodcast?.subscribedAt || syncedAt;
    const existingEpisodes = await getEpisodesForPodcast(normalizedUrl);
    const existingMap = new Map(existingEpisodes.map((episode) => [episode.episodeId, episode]));

    const podcast = {
      feedUrl: normalizedUrl,
      title: parsedFeed.title,
      author: parsedFeed.author,
      description: parsedFeed.description,
      imageUrl: parsedFeed.imageUrl,
      siteUrl: parsedFeed.siteUrl,
      subscribedAt,
      lastSyncedAt: syncedAt,
      notifyNewEpisodes: existingPodcast?.notifyNewEpisodes ?? true,
      includeInQueue: existingPodcast?.includeInQueue ?? true,
      introSkipSeconds: existingPodcast?.introSkipSeconds ?? 0,
      outroSkipSeconds: existingPodcast?.outroSkipSeconds ?? 0,
    };

    const episodes = [];
    const newEpisodeTitles = [];
    for (const parsedEpisode of parsedFeed.episodes) {
      const stableId = await rules.stableId(parsedEpisode);
      const merged = await rules.mergeWithExistingEpisode({
        parsedEpisode,
        podcastId: normalizedUrl,
        fallbackArtworkUrl: parsedFeed.imageUrl,
        existingEpisode: existingMap.get(stableId) || null,
      });
      episodes.push(merged);
      if (!existingMap.has(merged.episodeId)) {
        newEpisodeTitles.push(merged.title);
      }
    }

    await db.put("podcasts", podcast);
    await db.replacePodcastEpisodes(normalizedUrl, episodes);

    const shouldNotify = Boolean(notifyForNewEpisodes && podcast.notifyNewEpisodes && existingPodcast);
    return {
      feedUrl: normalizedUrl,
      podcast,
      newEpisodeTitles: shouldNotify ? newEpisodeTitles : [],
      shouldNotify,
    };
  }

  async function addSubscription(feedUrl) {
    const result = await syncFeed(feedUrl, false);
    emit();
    return result;
  }

  async function refreshAllFeeds() {
    const podcasts = await db.getAllFrom("podcasts");
    const summaries = [];
    for (const podcast of podcasts) {
      try {
        const result = await syncFeed(podcast.feedUrl, true);
        if (result.shouldNotify && result.newEpisodeTitles.length) {
          summaries.push(
            buildNewEpisodesSummary(
              result.feedUrl,
              result.podcast,
              result.newEpisodeTitles,
              result.shouldNotify
            )
          );
        }
      } catch (error) {
        console.warn("Skipping feed refresh after error", podcast.feedUrl, error);
      }
    }
    emit();
    return summaries;
  }

  async function getSubscriptionsSnapshot() {
    const podcasts = await db.getAllFrom("podcasts");
    const episodes = await db.getAllFrom("episodes");
    const countsByPodcast = new Map();
    for (const episode of episodes) {
      const current = countsByPodcast.get(episode.podcastId) || { episodeCount: 0, unreadCount: 0 };
      current.episodeCount += 1;
      if (!episode.isRead) {
        current.unreadCount += 1;
      }
      countsByPodcast.set(episode.podcastId, current);
    }
    return podcasts
      .map((podcast) => {
        const counts = countsByPodcast.get(podcast.feedUrl) || { episodeCount: 0, unreadCount: 0 };
        return {
          feedUrl: podcast.feedUrl,
          title: podcast.title,
          author: podcast.author,
          description: podcast.description,
          imageUrl: podcast.imageUrl,
          episodeCount: counts.episodeCount,
          unreadCount: counts.unreadCount,
          lastSyncedAt: podcast.lastSyncedAt,
          notifyNewEpisodes: podcast.notifyNewEpisodes,
          includeInQueue: podcast.includeInQueue,
          introSkipSeconds: podcast.introSkipSeconds,
          outroSkipSeconds: podcast.outroSkipSeconds,
        };
      })
      .sort((left, right) =>
        left.title.localeCompare(right.title, undefined, { sensitivity: "base" })
      );
  }

  async function getQueueSnapshot() {
    const [podcasts, episodes] = await Promise.all([db.getAllFrom("podcasts"), db.getAllFrom("episodes")]);
    const podcastsMap = new Map(podcasts.map((podcast) => [podcast.feedUrl, podcast]));
    return episodes
      .filter((episode) => podcastsMap.get(episode.podcastId)?.includeInQueue)
      .map((episode) => {
        const podcast = podcastsMap.get(episode.podcastId);
        return {
          ...episode,
          podcastTitle: podcast?.title || "",
          introSkipSeconds: podcast?.introSkipSeconds || 0,
          outroSkipSeconds: podcast?.outroSkipSeconds || 0,
        };
      })
      .sort(compareByPublishedAndTitleDesc);
  }

  async function getCalendarEpisodesSnapshot() {
    const [podcasts, episodes] = await Promise.all([db.getAllFrom("podcasts"), db.getAllFrom("episodes")]);
    const podcastsMap = new Map(podcasts.map((podcast) => [podcast.feedUrl, podcast]));
    return episodes
      .map((episode) => {
        const podcast = podcastsMap.get(episode.podcastId);
        return {
          episodeId: episode.episodeId,
          podcastId: episode.podcastId,
          podcastTitle: podcast?.title || "",
          title: episode.title,
          summary: episode.summary,
          artworkUrl: episode.artworkUrl,
          episodeUrl: episode.episodeUrl,
          publishedAt: episode.publishedAt,
          durationLabel: episode.durationLabel,
          audioUrl: episode.audioUrl,
          durationMs: episode.durationMs,
          playbackPositionMs: episode.playbackPositionMs,
          introSkipSeconds: podcast?.introSkipSeconds || 0,
          outroSkipSeconds: podcast?.outroSkipSeconds || 0,
          isRead: episode.isRead,
        };
      })
      .sort(compareByPublishedAndTitleDesc);
  }

  async function updatePlaybackState(episodeId, positionMs, durationMs, isRead, isCompleted) {
    const episode = await db.get("episodes", episodeId);
    if (!episode) {
      return;
    }
    await db.put("episodes", {
      ...episode,
      playbackPositionMs: Math.max(0, Number(positionMs) || 0),
      durationMs: durationMs > 0 ? durationMs : episode.durationMs,
      isRead: Boolean(isRead),
      isCompleted: Boolean(isCompleted),
    });
    emit();
  }

  async function markEpisodeRead(episodeId, isRead = true) {
    const episode = await db.get("episodes", episodeId);
    if (!episode) {
      return;
    }
    await db.put("episodes", {
      ...episode,
      isRead: Boolean(isRead),
      playbackPositionMs: isRead ? 0 : episode.playbackPositionMs,
    });
    emit();
  }

  async function markAllEpisodesReadForPodcast(feedUrl, isRead) {
    const episodes = await db.getManyByIndex("episodes", "by_podcast_id", feedUrl);
    const updated = episodes.map((episode) => ({
      ...episode,
      isRead: Boolean(isRead),
      playbackPositionMs: isRead ? 0 : episode.playbackPositionMs,
    }));
    await db.putMany("episodes", updated);
    emit();
  }

  async function updateSubscriptionSettings(
    feedUrl,
    notifyNewEpisodes,
    includeInQueue,
    introSkipSeconds,
    outroSkipSeconds
  ) {
    const podcast = await db.get("podcasts", feedUrl);
    if (!podcast) {
      return;
    }
    await db.put("podcasts", {
      ...podcast,
      notifyNewEpisodes: Boolean(notifyNewEpisodes),
      includeInQueue: Boolean(includeInQueue),
      introSkipSeconds: Math.max(0, Number(introSkipSeconds) || 0),
      outroSkipSeconds: Math.max(0, Number(outroSkipSeconds) || 0),
    });
    emit();
  }

  async function removeSubscription(feedUrl) {
    await db.deleteWhere("episodes", (episode) => episode.podcastId === feedUrl);
    await db.deleteKey("podcasts", feedUrl);
    emit();
  }

  async function getEpisodesForSubscription(feedUrl) {
    const episodes = await db.getManyByIndex("episodes", "by_podcast_id", feedUrl);
    return episodes.sort(compareByPublishedAndTitleDesc);
  }

  async function getEpisodeById(episodeId) {
    return db.get("episodes", episodeId);
  }

  return {
    subscribe,
    addSubscription,
    refreshAllFeeds,
    getSubscriptionsSnapshot,
    getQueueSnapshot,
    getCalendarEpisodesSnapshot,
    getEpisodesForSubscription,
    getEpisodeById,
    getPodcast,
    updatePlaybackState,
    markEpisodeRead,
    markAllEpisodesReadForPodcast,
    updateSubscriptionSettings,
    removeSubscription,
    compareByPublishedAndTitleDesc,
    compareByPublishedAndTitleAsc,
  };
}
